msdk-featdet/msdk-featdet-ADAP-3D/src/main/java/io/github/msdk/featdet/ADAP3D/common/algorithms/BiGaussian.java

```diff
@@ -0,0 +1,198 @@
+ *
+ * (a) the terms of the GNU Lesser General Public License version 2.1 as published by the Free
+ * Software Foundation
+ *
+ * or (per the licensee's choosing)
+ *
+ * (b) the terms of the Eclipse Public License v1.0 as published by the Eclipse Foundation.
+ */
+package io.github.msdk.featdet.ADAP3D.common.algorithms;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.lang.Math;
+
+import org.apache.commons.collections4.MapIterator;
+import org.apache.commons.collections4.map.MultiKeyMap;
+import org.apache.commons.collections4.keyvalue.MultiKey;
+
+/**
+ * <p>
+ * BiGaussian Class is used for fitting bigaussian on EIC.
+ * </p>
+ */
+public class BiGaussian {
+	
+	//This is used to store horizontal slice of sparse matrix.
+	@SuppressWarnings("rawtypes")
+	private final MultiKeyMap horizontalSlice;
+	
+	//This is used to store all the intensities of slice. 
+	private final List<Double> listOfIntensities;
+	
+	//This is used for storing bigaussian parameters inside constructor.
+	private final double biGaussianParams[];
+	
+	/**
+	 * <p>
+	 * Inside BiGaussian Constructor we're determining 4 BiGaussian Parameters. MaxHeight, Mu, SigmaLeft and SigmaRight.
+	 * </p>
+	 * 
+	 * @param horizontalSlice a {@link org.apache.commons.collections4.map.MultiKeyMap} object. This is horizontal slice from the sparse matrix.
+	 * @param mz  a {@link java.lang.Double} object. This is m/z value from the raw file.
+	 * @param leftBound a {@link java.lang.Integer} object. This is minimum scan number.
+	 * @param rightBound a {@link java.lang.Integer} object. This is maximum scan number.
+	 */
+	@SuppressWarnings({ "rawtypes", "unchecked" })
+	BiGaussian(MultiKeyMap horizontalSlice,double mz,int leftBound,int rightBound){
+		
+		this.horizontalSlice = horizontalSlice;
+		
+		List<SliceSparseMatrix.Triplet> listOfTripletsInSlice = new ArrayList<SliceSparseMatrix.Triplet>(horizontalSlice.values());
+		
+		listOfIntensities = new ArrayList<Double>();
+		
+		for(int i=0;i<listOfTripletsInSlice.size();i++){
+			double intensity = listOfTripletsInSlice.get(i)!=null?listOfTripletsInSlice.get(i).intensity:0;
+			listOfIntensities.add(intensity);
+		}
+		
+		
+		
+		Comparator<Double> compareIntensities = new Comparator<Double>() {
+			
+			@Override
+			public int compare(Double o1, Double o2) {
+				Double intensity1 = o1;
+				Double intensity2 = o2;
+				return intensity1.compareTo(intensity2);
+			}
+		};
+		
+		Collections.sort(listOfIntensities, compareIntensities);
+		
+		//This is max height for BiGaussian fit. It's in terms of intensities.
+		double maxHeight = listOfIntensities.get(listOfIntensities.size()-1);
+		
+		//Below logic is for finding BiGaussian parameters.		
+		int mu = getScanNumber(maxHeight);
+		int roundedmz = (int) Math.round(mz * 10000); 
+		double halfHeight = (double) maxHeight/2; 
+		biGaussianParams = new double[4];
```

Comment:
Creating a double[4] object is an awfully C++-like approach. Better to create a class to hold these parameters, and give them proper names. I mean, how are people supposed to know what is stored in biGaussianParams[2]?